torch/lib/THCUNN/SoftMaxCommon.cuh

```diff
@@ -0,0 +1,454 @@
+////////////////////////////////////////////////////////////////////////////////
+
+// Let's assume that our input has been flattened to have only three dimension:
+//     outer x dim x inner
+// The spatial algorithm tries to paralellize along all of them.
+// Within a 2d block threadIdx.y paralellizes over dim slices, and threads that
+// share it will speed up reductions over dim (along axis x).
+// The 2d grid is used to paralellize inner dimension over y axis and outer over x.
+
+inline dim3 SpatialSoftMax_getGridSize(
+    dim3 block, uint32_t max_active_blocks,
+    uint64_t outer_size, uint64_t dim_size, uint64_t inner_size) {
+  // First, tile as many blocks as we can over the y axis
+  uint32_t inner_blocks = (inner_size + block.y - 1) / block.y;
+  if (inner_blocks > max_active_blocks)
+    inner_blocks = max_active_blocks;
+  // Fill the x axis with as many blocks as we can fit (a little more is ok too)
+  uint32_t outer_blocks = (max_active_blocks + inner_blocks - 1) / inner_blocks;
+  if (outer_blocks > outer_size)
+    outer_blocks = outer_size;
+  return dim3(outer_blocks, inner_blocks);
+}
+
+inline dim3 SpatialSoftMax_getBlockSize(
+    uint64_t outer_size, uint64_t dim_size, uint64_t inner_size) {
+  uint32_t inner_threads = inner_size;
+  inner_threads = std::min(inner_threads, static_cast<uint32_t>(1024));
+  uint32_t dim_threads = 1;
+  if (inner_threads <= 64 && dim_size >= 64) {
+    while (inner_threads * dim_threads <= 1024 && dim_threads <= dim_size)
+      dim_threads *= 2;
+    dim_threads /= 2;
+  }
+  return dim3(dim_threads, inner_threads);
+}
+
+template<typename AccumT, typename Kernel>
+void SpatialSoftMax_getLaunchSizes(
+    THCState *state, Kernel k,
+    uint64_t outer_size, uint64_t dim_size, uint64_t inner_size,
+    dim3& grid, dim3& block, uint32_t& smem_size) {
+  block = SpatialSoftMax_getBlockSize(outer_size, dim_size, inner_size);
+  uint32_t block_threads = block.x * block.y;
+  smem_size = block.x == 1 ? 0 : block_threads * sizeof(AccumT);
+  int max_active_blocks;
+  cudaOccupancyMaxActiveBlocksPerMultiprocessor(&max_active_blocks,
+                                                k, block_threads, smem_size);
+  max_active_blocks *= THCState_getCurrentDeviceProperties(state)->multiProcessorCount;
+  grid = SpatialSoftMax_getGridSize(block, max_active_blocks, outer_size, dim_size, inner_size);
+}
+
+template<typename T>
+struct Add {
+  __device__ __forceinline__ T operator()(T a, T b) const {
+    return a + b;
+  }
+};
+
+template<typename T>
+struct Max {
+  __device__ __forceinline__ T operator()(T a, T b) const {
+    return a < b ? b : a;
+  }
+};
+
+// Note that it's not a complete block-wide reduction.
+// Only threads that share threadIdx.y reduce values.
+template<typename T, template<typename> class ReduceOp>
+__forceinline__ __device__
+T spatialBlockReduceX(T *shared, T val) {
+  ReduceOp<T> r;
+  shared += threadIdx.y * blockDim.x;
+
+  __syncthreads();
```

Comment:
Why is there a `__syncthreads()` here?

Reply:
It's just safer to have it. If you had two calls to `blockReduceX` one after another, it would be possible that some threads wouldn't read the return value in time before the threads that are already in the next call start writing their values

Reply:
I see. Does this actually occur in the code? It doesn't really matter because this function is limited to SoftMax but in general it seems like in the above case you would rather the calling code have to handle that if they planned to call it multiple times in a row, as opposed to make all callers pay the `__syncthreads()` cost.

Reply:
I just double checked and this instruction makes absolutely no difference in my benchmark. I'll leave it for safety

Reply:
Ok, my bad. For small inputs it gives 15% speedup and there's no difference for larger ones. Here's the plot (ignore the red dots for dim_size < 6, they are noise). @ngimel is it safe to remove this?